internal/provider/resource_copy.go

```diff
@@ -0,0 +1,178 @@
+	"github.com/hashicorp/terraform-plugin-framework/types"
+	"github.com/hashicorp/terraform-plugin-log/tflog"
+	"github.com/sigstore/cosign/v2/cmd/cosign/cli/copy"
+	"github.com/sigstore/cosign/v2/cmd/cosign/cli/options"
+)
+
+var (
+	_ resource.Resource                = &CopyResource{}
+	_ resource.ResourceWithImportState = &CopyResource{}
+)
+
+func NewCopyResource() resource.Resource {
+	return &CopyResource{}
+}
+
+type CopyResource struct {
+}
+
+type CopyResourceModel struct {
+	Id              types.String `tfsdk:"id"`
+	Source          types.String `tfsdk:"source"`
+	DestinationRepo types.String `tfsdk:"destination_repo"`
+
+	CopiedRef types.String `tfsdk:"copied_ref"`
+}
+
+func (r *CopyResource) Metadata(ctx context.Context, req resource.MetadataRequest, resp *resource.MetadataResponse) {
+	resp.TypeName = req.ProviderTypeName + "_copy"
+}
+
+func (r *CopyResource) Schema(ctx context.Context, req resource.SchemaRequest, resp *resource.SchemaResponse) {
+	resp.Schema = schema.Schema{
+		MarkdownDescription: "This copies the provided image digest cosign copy.",
+		Attributes: map[string]schema.Attribute{
+			"id": schema.StringAttribute{
+				Computed:            true,
+				MarkdownDescription: "The immutable digest this resource copies, along with its signatures, etc.",
+				PlanModifiers: []planmodifier.String{
+					stringplanmodifier.UseStateForUnknown(),
+				},
+			},
+			"source": schema.StringAttribute{
+				MarkdownDescription: "The digest of the container image to copy.",
+				Optional:            false,
+				Required:            true,
+				Validators:          []validator.String{validators.DigestValidator{}},
+				PlanModifiers: []planmodifier.String{
+					stringplanmodifier.RequiresReplace(),
+				},
+			},
+			"destination_repo": schema.StringAttribute{
+				MarkdownDescription: "The destination repository.",
+				Optional:            false,
+				Required:            true,
+				Validators:          []validator.String{validators.RepoValidator{}},
+				PlanModifiers: []planmodifier.String{
+					stringplanmodifier.RequiresReplace(),
+				},
+			},
+			"copied_ref": schema.StringAttribute{
+				MarkdownDescription: "This always matches the input digest, but is a convenience for composition.",
+				Computed:            true,
+			},
+		},
+	}
+}
+
+func (r *CopyResource) Configure(_ context.Context, req resource.ConfigureRequest, _ *resource.ConfigureResponse) {
+}
+
+func doCopy(ctx context.Context, data *CopyResourceModel) (string, error) {
+	digest, err := name.NewDigest(data.Source.ValueString())
+	if err != nil {
+		return "", errors.New("Unable to parse image digest")
+	}
+
+	ropts := options.RegistryOptions{
+		KubernetesKeychain: true,
+	}
+	dst, err := name.NewRepository(data.DestinationRepo.ValueString())
+	if err != nil {
+		return "", errors.New("Unable to parse destination repository")
+	}
+
+	if err := copy.CopyCmd(ctx, ropts, digest.String(), dst.String(), false, false); err != nil {
```

Comment:
I want to confirm that the semantics of this are to create the input reference in the `dst` repo, so if a tag is passed, the same tag is copied, but if a digest is passed the digest is copied (without tagging).

My concern here seeing the `dst.String()` is that the repo will be interpreted as having an implicit `:latest` and that we cannot use this without always overwriting `:latest`.

Reply:
If we made the target `dst.Digest(digest.DigestStr()).String()` that would be maximally clear, but not sure if `cosign` accepts that 😅

Reply:
Let's try it. I agree that `CopyCmd`'s args don't make it clear whether it takes a repo, a repo(-but-then-also-apply-`:latest`), or would accept a digest.